jlm/hls/backend/rvsdg2rhls/ThetaConversion.cpp

```diff
@@ -18,16 +18,31 @@ ConvertThetaNode(rvsdg::ThetaNode & theta)
   auto loop = hls::loop_node::create(theta.region());
   std::vector<jlm::rvsdg::input *> branches;
 
-  // add loopvars and populate the smap
+  // Add loop variables, insert loop constant buffers for invariant variables, and populate the
+  // smap.
   for (size_t i = 0; i < theta.ninputs(); i++)
   {
-    jlm::rvsdg::output * buffer;
-    loop->add_loopvar(theta.input(i)->origin(), &buffer);
-    smap.insert(theta.input(i)->argument(), buffer);
-    // buffer out is only used by branch
-    branches.push_back(*buffer->begin());
-    // divert theta outputs
-    theta.output(i)->divert_users(loop->output(i));
+    // Check if the input is a loop invariant such that a loop constant buffer should be created.
+    // Memory state inputs are not loop variables containting a value, so we ignor these.
+    if (is_invariant(theta.input(i))
+        && theta.input(i)->Type() != jlm::llvm::MemoryStateType::Create())
```

Comment:
Also, what about other subclasses of state types?

Reply:
There are only two subclasses of state types that should occur: IoStateType and MemoryStateType. IoStateTypes should not exist in HLS kernels, otherwise something went wrong (I believe).

Reply:
I agree with your first sentence though. I do not fully understand this either. @sjalander Could you elaborate, please?